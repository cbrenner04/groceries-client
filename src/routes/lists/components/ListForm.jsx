import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { SelectField, TextField } from '../../../components/FormFields';

function ListForm(props) {
  const defaultListType = 'GroceryList';
  const [name, setName] = useState('');
  const [type, setType] = useState(defaultListType);

  const handleSubmit = (event) => {
    event.preventDefault();
    props.onFormSubmit({
      name,
      type,
    });
    setName('');
    setType(defaultListType);
  };

  return (
    <form className="form" onSubmit={handleSubmit} autoComplete="off">
      <TextField
        name="name"
        label="Name"
        value={name}
        handleChange={({ target: { value } }) => setName(value)}
        placeholder="My super cool list"
      />
      <SelectField
        name="type"
        label="Type"
        value={type}
        handleChange={({ target: { value } }) => setType(value)}
        options={[
          { value: 'BookList', label: 'books' },
          { value: 'GroceryList', label: 'groceries' },
          { value: 'MusicList', label: 'music' },
          { value: 'ToDoList', label: 'to-do' },
        ]}
        blankOption={false}
      />
      <button type="submit" className="btn btn-success btn-block">Create List</button>
    </form>
  );
}

ListForm.propTypes = {
  onFormSubmit: PropTypes.func.isRequired,
};

export default ListForm;
