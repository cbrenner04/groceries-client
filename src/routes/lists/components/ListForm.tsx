import type { ChangeEvent, FormEvent } from 'react';
import React, { useState } from 'react';
import { Button, Collapse, Form } from 'react-bootstrap';

import ListFormFields from '../components/ListFormFields';
import FormSubmission from '../../../components/FormSubmission';
import type { IList } from '../../../typings';
import { EListType } from '../../../typings';

interface IListFormProps {
  onFormSubmit: (list: IList) => Promise<void>;
  pending: boolean;
}

const ListForm: React.FC<IListFormProps> = ({ onFormSubmit, pending }) => {
  const defaultListType = EListType.GROCERY_LIST;
  const [name, setName] = useState('');
  const [type, setType] = useState(defaultListType as EListType);
  const [showForm, setShowForm] = useState(false);

  const handleSubmit = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    // TODO: figure this out.
    await onFormSubmit({ name, type } as IList);
    setName('');
    setType(defaultListType);
  };

  return (
    <>
      {!showForm && (
        <Button variant="link" onClick={() => setShowForm(true)} aria-controls="form-collapse" aria-expanded={showForm}>
          Add List
        </Button>
      )}
      <Collapse in={showForm}>
        <Form id="form-collapse" onSubmit={handleSubmit} autoComplete="off">
          <ListFormFields
            name={name}
            type={type}
            handleNameChange={({ target: { value } }: ChangeEvent<HTMLInputElement>) => setName(value)}
            handleTypeChange={({ target: { value } }: ChangeEvent<HTMLInputElement>) => setType(value as EListType)}
          />
          <FormSubmission
            disabled={pending}
            submitText="Create List"
            cancelAction={() => setShowForm(false)}
            cancelText="Collapse Form"
            displayCancelButton={true}
          />
        </Form>
      </Collapse>
    </>
  );
};

export default ListForm;
