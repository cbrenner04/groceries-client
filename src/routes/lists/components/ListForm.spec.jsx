import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';

import ListForm from './ListForm';

describe('ListForm', () => {
  it('renders', () => {
    const { container } = render(<ListForm pending={false} onFormSubmit={jest.fn()} />);

    expect(container).toMatchSnapshot();
  });

  it('expands form', async () => {
    const { baseElement, getByText } = render(<ListForm pending={false} onFormSubmit={jest.fn()} />);

    fireEvent.click(getByText('Add List'));
    await waitFor(() => expect(baseElement.children[0].children[0]).toHaveClass('show'));

    expect(baseElement.children[0].children[0]).toHaveClass('show');
  });

  it('collapses form', async () => {
    const { baseElement, getByText } = render(<ListForm pending={false} onFormSubmit={jest.fn()} />);

    fireEvent.click(getByText('Add List'));
    await waitFor(() => expect(baseElement.children[0].children[0]).toHaveClass('show'));

    fireEvent.click(getByText('Collapse Form'));
    await waitFor(() => expect(baseElement.children[0].children[0]).not.toHaveClass('show'));

    expect(baseElement.children[0].children[0]).not.toHaveClass('show');
  });

  it('changes the value in the name field', () => {
    const { getByLabelText } = render(<ListForm pending={false} onFormSubmit={jest.fn()} />);

    fireEvent.change(getByLabelText('Name'), { target: { value: 'foo' } });

    expect(getByLabelText('Name')).toHaveValue('foo');
  });

  it('changes the value in the type field', () => {
    const { getByLabelText } = render(<ListForm pending={false} onFormSubmit={jest.fn()} />);

    fireEvent.change(getByLabelText('Type'), { target: { value: 'MusicList' } });

    expect(getByLabelText('Type')).toHaveValue('MusicList');
  });

  it('calls props.onFormSubmit when form is submitted', async () => {
    const onFormSubmit = jest.fn().mockResolvedValue({});
    const { getByLabelText, getAllByRole } = render(<ListForm pending={false} onFormSubmit={onFormSubmit} />);

    fireEvent.change(getByLabelText('Name'), { target: { value: 'foo' } });
    fireEvent.change(getByLabelText('Type'), { target: { value: 'BookList' } });
    fireEvent.click(getAllByRole('button')[1]);

    await waitFor(() => expect(onFormSubmit).toHaveBeenCalledTimes(1));

    expect(onFormSubmit).toHaveBeenCalledWith({ name: 'foo', type: 'BookList' });
  });

  it('disables submit when in pending state', () => {
    const onFormSubmit = jest.fn().mockResolvedValue({});
    const { getAllByRole } = render(<ListForm pending={true} onFormSubmit={onFormSubmit} />);

    expect(getAllByRole('button')[1]).toBeDisabled();
  });
});
