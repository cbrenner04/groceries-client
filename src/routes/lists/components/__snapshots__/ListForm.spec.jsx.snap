// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListForm renders 1`] = `
<div>
  <button
    aria-controls="form-collapse"
    aria-expanded="false"
    class="btn btn-link"
    type="button"
  >
    Add List
  </button>
  <form
    autocomplete="off"
    class="collapse"
    id="form-collapse"
    style="padding: 6px;"
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="name"
      >
        Name
      </label>
      <input
        class="form-control"
        id="name"
        name="name"
        placeholder="My super cool list"
        type="text"
        value=""
      />
      
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="type"
      >
        Type
      </label>
      <select
        class="form-control"
        id="type"
        name="type"
      >
        <option
          value="BookList"
        >
          books
        </option>
        <option
          value="GroceryList"
        >
          groceries
        </option>
        <option
          value="MusicList"
        >
          music
        </option>
        <option
          value="SimpleList"
        >
          simple
        </option>
        <option
          value="ToDoList"
        >
          to-do
        </option>
      </select>
      
    </div>
    <button
      class="btn btn-success btn-block"
      type="submit"
    >
      Create List
    </button>
    <button
      class="btn btn-link btn-block"
      type="button"
    >
      Collapse Form
    </button>
  </form>
</div>
`;
